FAKE 5

matthid opened the pull request:
From https://github.com/fsharp/FAKE/issues/1232
## Docs

https://github.com/matthid/FAKE/blob/coreclr/help/fake-dotnetcore.md
## To test this branch

```
./build.sh
```
## TODO (for alpha release):
- [x] bootstrap the build process
- [x] https://github.com/dotnet/coreclr/blob/master/Documentation/building/debugging-instructions.md -> Workaround implemented
- [x] Mono.Cecil (https://github.com/jbevain/cecil/pull/269) 
- [x] Argu (https://github.com/fsprojects/Argu/pull/57)
- [x] Make Fake work with `dotnet run -f net46` (and remove everything but the core)
- [x] Make Fake run with `dotnet run -f netcoreapp1.0`
- [x] Things to cleanup
  - [X] FCS.netcore: Chockes on `checker.ParseAndCheckProject(options)`: `internal error: Object reference not set to an instance of an object.` (Workaround for now -> no `fsi` object in build script). Disappeared after latest FCS update (merge with visualfsharp)
  - [x] `Yaaf.FSharp.Scripting`: How do we resolve `FSharp.Core` with `sigdata` and `optdata` in this new world? Required for `let` functionality. (Workaround for now -> no `fsi` object in build script)
  - [x] ~~Error `**** error: internal error: The type initializer for '<StartupCode$FSharp-Compiler-Service-netcore>.$Reshapedmsbuild' threw an exception.` when running with `fsi` object support~~ -> not important in the first version
  - [x] `error FS0192: internal error: the IL instruction Microsoft.FSharp.Compiler.AbstractIL.IL+ILInstr+I_seqpoint cannot be emitted`, See https://github.com/fsharp/FSharp.Compiler.Service/issues/591 and https://github.com/fsharp/FSharp.Compiler.Service/pull/592
  - [x] Remove hardcoded references in `FSIHelper.fs`
  - [x] Remove hardcoded path in `Yaaf.FSharp.Scripting`
  - [x] ~~Remove binaries from lib folder (currently these "special" version are required)~~ -> Can be done one by one once they become available.
- [x] Try to make it run standalone -> we can test the bootstrapping experience from this point on (but FAKE scripts cannot have dependencies)
- [x] ~~Add the `fake` coreclr tool -> maybe bootstrapping is better?~~ -> Later
- [x] Make Paket CoreCLR ready (https://github.com/fsprojects/Paket/pull/1785)
- [x] Add Paket.Core to FAKE, read the script, add the resolving step and add the `#load` lines
  - [x] Restore after Update as well
  - [x] Bug: Relative references via "paket.inline"
  - [x] ~~Make the caching layer in fake load the correct assembly (depending on the current runtime and the runtime folder of the packages) if necessary (Microsoft.NETCore.Platforms)~~ -> Not a showstopper for now as for Fake-Packages runtime- and compile-time assemblies are the same. We fallback to already loaded assemblies anyway (therefore we can load netstandard for now)
- [x] Move the new "Core" to "Fake.Runtime"
  - [x] ~~Remove duplicate code from FakeLib (mutables!)~~ Will be done on demand.
  - [x] Remove Conditional Compilation from LegacyApi (was the first approach)
  - [x] Remove StringHelpers from Fake.Runtime
  - [x] Remove BuildServerHelper from Fake.Runtime (the interface probably need to remain, but it should be possible to extend)
  - [x] Strip FSIHelper.
- [x] Bootstrapping
- [x] Testing
  - [x] Add dotnetcore build to AppVeyor
  - [x] and Travis
  - [x] Add integration tests for the dotnetcore build
- [x] 'New' Library / Fake Components (a lot of packages are out of scope for this PR, see https://github.com/fsharp/FAKE/issues/1232)
  - [x] Fake.Core.Environment
  - [x] Fake.Core.String
  - [x] Fake.Core.BuildServer
  - [x] Fake.Core.Tracing
  - [x] Fake.Core.Targets
  - [x] Fake.Core.Tasks
  - [x] Fake.DotNet.MsBuild
  - [x] Fake.IO.FileSystem
  - [x] Fake.Core.Globbing
  - [x] Fake.Core.Process
  - [x] Fake.DotNet.Cli
  - [x] Fake.Testing.Common
  - [x] Fake.DotNet.Testing.MSpec
  - [x] Fake.DotNet.Testing.NUnit
  - [x] Fake.DotNet.Testing.XUnit2
  - [x] Fake.Core.SemVer
  - [x] Fake.Windows.Chocolatey
  - [x] Fake.Core.ReleaseNotes
  - [x] Fake.DotNet.NUnit
  - [x] ~~Fake.BuildServer.AppVeyor~~ -> Later
  - [x] ~~Fake.DotNet.FSharpFormatting~~ -> Later
  - [x] Fake.DotNet.AssemblyInfo
  - [x] Fake.Tracing.NAntXml
- [x] Fix Publish on preview4.
- [x] Future: ~~Trim packages ([here](https://docs.microsoft.com/de-de/dotnet/articles/core/deploying/reducing-dependencies) and [here](https://docs.microsoft.com/de-de/dotnet/articles/core/tutorials/managing-package-dependency-versions))~~, see https://github.com/fsharp/FAKE/issues/1232

matthid commented:
@forki FYI
```
$ ./build.sh Default

Starting download from https://github.com/fsprojects/Paket/releases/download/3.3.2/paket.exe

Done.

Paket version 3.3.2.0

Paket failed with:

	Error in paket.dependencies line 33

  Package xunit.runners is defined more than once in group Main of /home/travis/build/fsharp/FAKE/paket.dependencies
```

cloudRoutine commented:
if you're still having build issues try it with these settings in your project.json
```json
{
  "name": "",
  "version": "1.0.0-*",
  "buildOptions": {
    "compilerName": "fsc",
    "compile": {
      "includeFiles": [

      ]
    }
  },
  "dependencies": {
    "Microsoft.FSharp.Core.netcore": "1.0.0-alpha-160509",
    "FSharp.Compiler.Service.netcore": "1.0.0-alpha-00001",
    "FSharp.Compiler.Service.ProjectCracker.netcore": "1.0.0-alpha-00001",
    "NETStandard.Library": "1.5.0-rc2-24027"
  },
  "tools": {
    "dotnet-compile-fsc": {
      "version": "1.0.0-*",
      "imports": [
        "dnxcore50",
        "portable-net45+win81",
        "netstandard1.3"
      ]
    }
  },
  "frameworks": {
    "netstandard1.5": {
      "imports": [
        "portable-net45+win8",
        "dnxcore50"
      ]
    }
  }
}
```

and these in your nuget.config
```xml
<?xml version="1.0" encoding="utf-8"?>
<configuration>
  <packageSources>
    <!--To inherit the global NuGet package sources remove the <clear/> line below -->
    <clear />
    <add key="dotnet-core" value="https://dotnet.myget.org/F/dotnet-core/api/v3/index.json" />
    <add key="dotnet-cli" value="https://dotnet.myget.org/F/dotnet-cli/api/v3/index.json" />
    <add key="cli-deps" value="https://dotnet.myget.org/F/cli-deps/api/v3/index.json"/>   
    <add key="fsharp-daily" value="https://www.myget.org/F/fsharp-daily/api/v3/index.json" />    
    <add key="fsharp-compiler-service" value="https://ci.appveyor.com/nuget/fsgit-fsharp-compiler-service"/>
    <add key="api.nuget.org" value="https://api.nuget.org/v3/index.json" />
  </packageSources>
</configuration>
```

matthid commented:
@cloudRoutine  Build is fine net46 and netcoreapp1.0 (the AppVeyor/Travis issue is unrelated). Running in net46 works as well. Only running Fake.netcore with netcoreapp1.0 is broken. I think I have most of the stuff already added to project.json and NuGet.config, but I will look again later, thanks.

matthid commented:
Updated first post with the current errors I'm encountering...
Seems like I build FSharp.Compiler.Service wrong?
IIRC I did just `build DotnetCliBuild` and then copied the resulting nuget package `FSharp.Compiler.Service.netcore.1.0.0-alpha-00001.nupkg` to this branch.

@cloudRoutine I switched to the myget feed which contains `FSharp.Compiler.Service` (and deleted local nuget caches). Sadly the same thing happens. Thanks again for the tipp.

cloudRoutine commented:
@matthid  I managed to get it to run, but I had to remove some dependencies and functionality. You'll need to find compatible versions to work with what's here or try another implementation approach.

[Fake.netcore](https://gist.github.com/cloudRoutine/0006bf3c79626203c0d32a52ef76d999)
[Fakelib.netcore](https://gist.github.com/cloudRoutine/bc012df5574b401128b5970fff83a4c1)
![](http://i.imgur.com/cdpQWLc.png)

matthid commented:
@cloudRoutine with dotnet version are you using?
```
$ dotnet --version
1.0.0-preview1-002702
```

It's already working for me (at least after running `dotnet publish`, see first post). It just seems there are some bugs in `FSharp.Compiler.Service.netcore`

cloudRoutine commented:
```
dotnet --version
1.0.0-preview3-003161
```
i guess you're on top of it then :+1:

matthid commented:
I have no idea. This whole versioning is so confusing. I have no idea if I should upgrade and try do make it work with preview3 or stay on preview1 which is suggested here: https://github.com/dotnet/netcorecli-fsc/wiki

cloudRoutine commented:
Honestly I'd wait until everything becomes more stable. There's no huge gain in rushing to try to get it done early when it could be done later much faster with less headaches.

matthid commented:
Wow... Now I get some internal F#  compiler error: `error FS0192: internal error: the IL instruction Microsoft.FSharp.Compiler.AbstractIL.IL+ILInstr+I_seqpoint cannot be emitted`.

I think I have to give up for now.

@cloudRoutine Yeah I kind of agree for the cli tooling, that's why I have given up on the `run -f netcoreapp1.0` thing. But as it "works" after publishing and the runtime should be "stable" now I think it is the time to make FAKE and the FSharp.Compiler.Service ready :).

Most of the errors now are probably FCS specific bugs and I'm not sure anyone has interest in fixing them (if not FAKE).

matthid commented:
Related: https://github.com/fsharp/FSharp.Compiler.Service/pull/593 (you need those changes for this branch)

matthid commented:
Wow I made it work:
```
$ cat test.fsx
printfn "Hello FAKE"
```

```
$ dotnet exec bin/Debug/netcoreapp1.0/publish/Fake.netcore.dll run --verbose test.fsx
FakePath: C:\PROJ\FAKE\src\app\Fake.netcore\bin\Debug\netcoreapp1.0\publish\FakeLib.netcore.dll
FAKE - F# Make "4.29.2"
Running Buildscript: test.fsx
Cache doesn't exist
FAKE: Trying to resolve FSharp.Compiler.Service.netcore.resources, Version=1.0.0.0, Culture=en-US, PublicKeyToken=null
FAKE: Trying to resolve FSharp.Compiler.Service.netcore.resources, Version=1.0.0.0, Culture=en, PublicKeyToken=null
FAKE: Trying to resolve System.Core, Version=3.5.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089
FAKE: Trying to resolve FSharp.Compiler.Service.netcore.resources, Version=1.0.0.0, Culture=en-US, PublicKeyToken=null
FAKE: Trying to resolve FSharp.Compiler.Service.netcore.resources, Version=1.0.0.0, Culture=en, PublicKeyToken=null
Hello FAKE
Caching is disabled on CoreCLR, see https://github.com/dotnet/coreclr/issues/919#issuecomment-219212910
Ready.
```

matthid commented:
Updated first post, updated to latest preview 2 tooling and changed to Argu for command line parsing.

matthid commented:
I made a first release for win7 to check how bootstrapping could look like: https://github.com/matthid/FAKE/releases/tag/v1.0-alpha-01

matthid commented:
How I'm moving this forward and my suggested migration path:

 - Creating a new library "FakeLib.Core" which will be the absolute core of what we need to run scripts
 - Move existing functionality one by one to new packages and make them reference each other and "FakeLib.Core" (if they need it). The new packages should be usable as libraries if possible (from regular Programs).
 - The existing "FakeLib" should reference all the new packages and "FakeLib.Core". This way we can continue delivering the known "Fake" (the package will just contain more files). We just need to make sure that `#r "FakeLib.dll"` still loads all files. (needs to be tested, if not we can create the "old" FakeLib by using ILMerge)
 - Once everything is converted we could in theory get rid of "FakeLib" and write a new console application using the new packages, while providing the "old" command-line interface. This would print a warning to the user to update Fake to the dotnetcore version.

This is the general Roadmap. For now I'm concentrating on the bootstrap experience and to get the most important features converted to separate packages.

The road so far:
 - Released https://github.com/matthid/FAKE/releases/tag/core-v1.0-alpha-02
 - Run simple scripts without dependencies
 - Fake is now capable of restoring dependencies before running the build script. You can specify dependencies within your build script in a `Fake Dependencies` section.
 - Either embed your dependencies, like:

   ```
(* -- Fake Dependencies paket-inline
source https://nuget.org/api/v2

nuget System.Xml.XmlDocument ~> 4.0
nuget System.Security.Cryptography.Algorithms ~> 4.0
nuget Microsoft.FSharp.Core.netcore
-- Fake Dependencies -- *)
#load "./.fake/unusedDependencies.fsx/loadDependencies.fsx"

open System.Xml

printfn "Starting Build."
```

 - Or use an external `paket.dependencies` file:

    ```
(* -- Fake Dependencies paket.depedencies
file ./paket.dependencies
group Main
-- Fake Dependencies -- *)
#load "./.fake/use_external_dependencies.fsx/loadDependencies.fsx"


printfn "Starting Build."
```

   Note the typo "paket.depedencies" in the header. This will be fixed in the next version. Note also the "./" in the file specifier, this is currently required as well.

 - The command line is using Argu 3 and SubCommands (`fake.netcore.exe run` is for running scripts, just look into the `--help` output). I'm thinking about adding more sub-commands in the future.
 - The `loadDependencies.fsx` file is currently empty. It will be used to provide nice editor support for editing scripts in the future (it is not required/use by fake).

smoothdeveloper commented:
Somehow related a set of helpers to run dotnet.exe commands from FAKE:

https://github.com/dolly22/fake.dotnet

I'm also having a wild idea: since there are so many FAKE extensions (in separate packages), maybe we could study a ExtensionTypeProvider which takes care of downloading the package and exposing the helpers to the FAKE Script?

matthid commented:
Workaround for https://github.com/fsprojects/Paket/issues/1815 and rebased on latest master

@smoothdeveloper 

> https://github.com/dolly22/fake.dotnet

Thanks will need that soon :)

> I'm also having a wild idea: since there are so many FAKE extensions (in separate packages), maybe we could study a ExtensionTypeProvider which takes care of downloading the package and exposing the helpers to the FAKE Script?

There will be a lot extensions soon :). I have no idea how that could work as the type provider is working at compile time (when you write the script). But then the type provider needs to add the information somewhere such that the extension can be found at runtime as well. Maybe the TP could update the paket.dependencies file, but no idea how well that would work in practice. Interesting idea though (I think one can experiment with this on top of these changes).

cloudRoutine commented:
@smoothdeveloper an extension/component type provider is like asking for compatibility issues and builds that are hard to reproduce unless it includes a lock system like paket and the ability to use caches. it wouldn't be able to provide access to the full scope of the API inside it if you're talking about trying to feed the code through the provider, and if you're not doing that what's the advantage of having it in the first place?  

The idea doesn't totally make sense or fit with what type providers are supposed to be used for.

matthid commented:
In theory we now have Targets support:
https://github.com/matthid/FAKE/releases/tag/core-v1.0-alpha-03

```fsharp
(* -- Fake Dependencies paket-inline
source https://nuget.org/api/v2
source .\..\..\..\nuget\dotnetcore

nuget Fake.Core.Targets prerelease strategy: min
nuget Microsoft.FSharp.Core.netcore strategy: min
-- Fake Dependencies -- *)
#load "./.fake/reference_fake-runtime.fsx/loadDependencies.fsx"

open Fake.Core
open Fake.Core.Targets
open Fake.Core.TargetOperators

Target "Start" (fun _ -> ())

Target "TestTarget" (fun _ ->
    printfn "Starting Build."
    Trace.traceFAKE "Some Info from FAKE"
    printfn "Ending Build."
)

"Start"
  ==> "TestTarget"

RunTargetOrDefault "TestTarget"
```

Seems to work but dotnetcore is crashing with a segfault :)

matthid commented:
Ok Targets are working now! The dotnetcore team helped me out really quickly and linked a workaround.

![working_targets](https://cloud.githubusercontent.com/assets/1866463/17066935/a85cc9dc-5049-11e6-8b5b-402bceab0f7f.jpg)

smoothdeveloper commented:
> The dotnetcore team helped me out really quickly and linked a workaround.

Great! also I'm sure some of the members are really eager to get their hands on FAKE to make builds the same breeze it is on .NET Framework / Mono :smile: 

You are a hero for bringing FAKE to dotnet core!

matthid commented:
@smoothdeveloper Thanks! FAKE will look slightly different after this, so lets see if people actually like it.
The only thing I will add is `Fake.Process` because that one is not trivial, rest will be a community effort and this PR can be merged. I think I can add issues for all the open points and do them one by one.

@forki Are you OK with moving a lot of files around and marking a lot of stuff obsolete (https://ci.appveyor.com/project/SteffenForkmann/fake/build/1.0.2477#L6097)? This has some advantages after this is finished.

matthid commented:
Ok I just released [Apha 4](https://github.com/matthid/FAKE/releases/tag/core-v1.0-alpha-04) of this. 

There is even some [documentation](https://github.com/matthid/FAKE/blob/coreclr/help/fake-dotnetcore.md) now.

We can even build [fake itself](https://github.com/matthid/FAKE/blob/coreclr/dncbuild.fsx) with this (msbuild, dotnet calls,...). So it's slowly getting usable.

matthid commented:
@forki Do you see any reason currently to not merge this / showstopper (NOTE: I'm not finished jet but if there are things I should improve I'd like to know early)

Because the diff is too large, here a list of (high level) changes:
- Moved all files in "FakeLib" to "LegacyApi" (and marked a lot as obsolete)
- Componentized "FakeLib" into smaller packages, but added the new API to "FakeLib" as well (-> therefore marked to old as obsolete)
- Added "FakeRuntime" (handles the paket interaction) and cleaned up "FsiHelpers.fs" for dotnetcore
- Added Integration tests for the new dotnetcore stuff
- Some Cleanup in the "new" moved API itself -> moved some stuff to different modules.
  Sadly git is not showing diffs there and just assumes new files (when in fact most is just copied and modified). I did this to be sure that I don't break old stuff (left most of the old stuff untouched, just moved it).

forki commented:
I think my first question is still: how many projects will we break?
I think we definetly should not break ProjectScaffold like projetcs.
Also I'd like to avoid breaking old style command line syntax, but that's not completely required.

matthid commented:
Currently we don't break anything (at least no breakage is intended with this PR) but write a lot of obsolete warnings. The new CLI will be side by side and distributed differently and I'm not sure jet if we can ever completely replace it...

forki commented:
ok good. but why exactly make are we making things obsolete?

matthid commented:
Because I don't want to split Fake into two APIs and support them both... I first tried to mixin the support for dotnetcore, but it wasn't easily possible and its just too much work for a single PR (see my first commits). After that I decided to componetize Fake so it better fits the new dotnetcore world. As a bonus we can now use the Fake components are regular libraries in regular projects without pulling the whole FakeLib and all of its dependencies.

matthid commented:
Currently blocking for proper Ubuntu support (travis): https://github.com/fsharp/FSharp.Compiler.Service/issues/619

Yantrio commented:
Is there anything I can do to help with this? I want to be 100% away from mono ASAP.

matthid commented:
Getting rid of mono completely is a long road... Most tools you would use in combination with FAKE still depend on mono one way or another (test-runner, xbuild, ...).

I consider the current code base as "reasonable". 
What needs to be done to make a "proper" release:

- [ ] Some dependencies need to be moved forward, like Paket, Argu, Mono.Cecil, FSharp.Compiler.Service and FSharp.Core. Only if all dependencies have a proper dotnetcore release you can expect a proper dotnetcore release from FAKE. I'd say the coding work is mostly finished, only some release processes need to be adjusted and releases need to be done. 
- [ ] The release process here needs to be adjusted to upload all the new packages.
- [ ] API needs to stabilize as there are a lot of changes.

An easy place to start would be to just use the current releases and play around with them (They are called alpha for a reason though: https://github.com/matthid/FAKE/releases/tag/core-v1.0-alpha-08).

Linux/OSx support should work. The build is red, because the dotnetcore helpers are currently not working on Linux/OSx (https://github.com/matthid/FAKE/blob/coreclr/src/app/Fake.DotNet.Cli/Dotnet.fs)

If someone is willing to help just mention me on gitter and say which area you wanna improve, such that we don't do the same thing multiple times :+1: . Reporting problems/bugs in the new API is always welcome as well.

matthid commented:
Note that the current release has the following limitation on Linux/OSx:
The environment variables `APPDATA` and `LocalAppData` need to be set. You can use
```bash
export APPDATA=$HOME LocalAppData=$HOME # (make sure HOME is set as well)
```
See https://github.com/matthid/FAKE/blob/coreclr/.travis.yml#L6

matthid commented:
For completeness sake, another related bug (until we don't need a local repository anymore -> everything is uploaded to nuget.org): https://github.com/NuGet/Home/issues/1672

matthid commented:
@Yantrio Can you play with alpha 10 and see if it works for you (https://github.com/matthid/FAKE/releases/tag/core-v1.0-alpha-10).
The "real" alpha or beta release will come later, when all deps are actually available on nuget....
For now the bootstrapping script is downloading 'fake-dotnetcore-packages.zip' which contains all the packages we need for now (just use `source .fake\bin\core-v1.0-alpha-10\packages`).

matthid commented:
See https://github.com/dotnet/coreclr/issues/7043

matthid commented:
See https://github.com/dotnet/coreclr/issues/7044

enricosada commented:
@matthid the `dotnet` path resolution doesnt work i think.
It should use always the `dotnet` in `PATH`, that's because how things are expect to work.
It's possibile to run the dotnet process as shell? Or get the location of dotnet with something similar to `where`/`which`?

an example: https://github.com/fsprojects/Paket/pull/1914

travis can install .net core sdk using `dotnet` key, so it's easy to have it installed by default.

a normal install:

- install .net core sdk to a specific directory
- add that directory to `PATH`
- `dotnet` now works.

enricosada commented on src/app/Fake.DotNet.Cli/Dotnet.fs:
```diff
@@ -0,0 +1,582 @@
+/// ## Parameters
+///
+/// - 'dotnetCliDir' - dotnet cli install directory 
+let private dotnetCliPath dotnetCliDir = dotnetCliDir @@ (if Environment.isUnix then "dotnet" else "dotnet.exe")
```
we should search in `PATH`

```
    match tryFindFileOnPath (if isWindows then "dotnet.exe" else "dotnet") with
    | Some p -> p
    | None -> ""
```

enricosada commented on src/app/Fake.DotNet.Cli/Dotnet.fs:
```diff
@@ -0,0 +1,582 @@
+let mutable DefaultDotnetCliDir = 
+    if Environment.isUnix
+    then Environment.environVar "HOME" @@ ".dotnet"
+    else Environment.environVar "LocalAppData" @@ "Microsoft" @@ "dotnet"
```
that's not ok, location should be resolved from `PATH`, there is no default install

matthid commented:
@enricosada I'm not sure if I agree. First time we can have reproducable builds independent from the installed .net version by just downloading our own dotnet sdk. 

At the very least I'd like to have the option to do so. To fallback to dotnet on PATH if nothing else is specified that might make sense.

Also it's about adoption. I'd very much prefer a solution where people don't need to install .net sdk to be able for any F# project to clone and build.

'./fake run InstallDotNetSdk' should probably not change user path, because I don't expect it to permanently change anything. 

Does that sound reasonable?

matthid commented:
I'm also not sure if 'Microsoft is suggesting this method' is enough of a reason to follow a particular strategy. If it would we probably should be using C# ;)

This PR tries to take the new portable clr (rtm) to its full advantage (ie remove the need for an installed runtime) and at the same time tries to minimize the use/need of the new tooling, which is still preview.

Once the tooling is matured we might reconsider this. Now its more of a neccessary evil. Best case is to generate 'project.json' from 'fsproj' and completely hide it from the user... (IMHO)

matthid commented:
@enricosada TLDR: I agree that the defaults maybe should be changed...

enricosada commented:
@matthid fsproj => project.json is not needed, `project.json` it's going to die, so np.

I agree is nice to have in fake the possibility to install the .net core sdk, and *optional* install of deps is nice

But there are some issue in the current implementation of fake helpers we should address, not a big deal.
I can send a pr, but i'd like to discuss that a bit before, because i dont know a lot of FAKE internal, user expectaction, conventions
Sry if it's long

## default location

check in `PATH` the default location of `dotnet`, no magic path. Np about that.

## use current one if already installed

if `dotnet --version` return the required version, np need to install. 
i dont know if it's already like that, but should, to avoid pain (slow download, etc)
Otherwise is like vf# repo after a git clean.

## the dotnet in PATH rules

the .net core sdk expect dotnet in `PATH`. That's how it works, we cannot just invoke the right `dotnet` executable to change the version used.
There was a `DOTNET_HOME` env var (like `JAVA_HOME`) but was removed in preview1, because `PATH` was preferred and now tooling expect that behaviour.
An example are build events like `postbuild` etc, who can run **arbitrary** shell command, who inherits environment variable from the process, but if `dotnet` is not in `PATH` (or the wrong one) there will be issues.
Another example are tools, who can invoke `dotnet` from shell.execute, because they dont need to know where is.

If you want to change the current dotnet just for the execution of the fake script:

- use `Environment.SetEnvironmentVariable` to change `PATH` of current project (fake script). It's scoped on current process, so doent mess with shell/global of user.
- run always the `dotnet` in `PATH`, or shell.execute `dotnet`

## behaviour on dotnet not found (something i'd like)

That's just a suggestion, make possibile to choose behaviour of fake when dotnet doesnt exists.

- fails with a message (dotnet not found or wrong version, install it or add autoinstall to fake script)
- silently install it
- a fake property from command line to choose the specific install directory (so i can do that outside of the repo dir, so `git clean` doesnt remove it), and if needed, is installed there

matthid commented:
@enricosada 

> fsproj => project.json is not needed, project.json it's going to die, so np.

But for now its a pain point. If we can get rid of it for now with minimal overhead that would be nice (do we already have parsers for fsproj somewhere? Maybe in Forge?). It doesn't need to be a perfect solution. Fake only needs to:
 
 - Edit Version (as we don't want to update it on two places)
 - Edit Files (as we don't wan't to add a file twice)
 - (Maybe) edit dependencies -> Version from lockfile?

This shouldn't be too hard to do. Rest of the `project.json` file can be a simple template.

>  is like vf# repo after a git clean

I don't mind a lot of downloads. What really is a mess at vf# are the number of global install requirements. I never compiled all the code base, because I don't like to install all this stuff on my machine.

> the .net core sdk expect dotnet in PATH. That's how it works, we cannot just invoke the right dotnet executable to change the version used.

Good to know. That's what we do right now and apparently its working anyway. But fixing that shouldn't be a problem.

> make possibile to choose behaviour of fake when dotnet doesn't exists.

You mean doesn't exist OR has the wrong version? Than I'd agree.

> silently install it

Yes, like paket we could cache the install in the user folder to share it across projects?

> a fake property from command line to choose the specific install directory (so i can do that outside of the repo dir, so git clean doesnt remove it), and if needed, is installed there

Do we need that if we always write into the user folder (maybe with the version in the path, which shouldn't be too bad as it sounds, as anyone can find it in `PATH`)?

matthid commented:
Now "FSharp.Compiler.Service" is the last dependency missing on nuget...

matthid commented:
Latest build errors are related to https://github.com/fsharp/fsharp/issues/608

matthid commented:
@enricosada Currently `dotnet publish` is failing when trying to build for a specific RID (I assume the example is https://github.com/dotnet/netcorecli-fsc/tree/master/examples/preview3/console-crossgen).

```
C:\PROJ\FAKE\test>C:\Users\dragon\AppData\Local\Microsoft\dotnet\dotnet.exe publish C:\PROJ\FAKE\src\app\Fake.netcore\Fake.netcore.fsproj --configuration Release --framework "netcoreapp1.0" --runtime "win10-x64" --output "C:\PROJ\FAKE\nuget\dotnetcore\Fake.netcore\win10-x64"
Microsoft (R) Build Engine version 15.1.458.808
Copyright (C) Microsoft Corporation. All rights reserved.

error FS0193 : internal error : The type initializer for '<StartupCode$FSharp-Compiler>.$Reshapedmsbuild' threw an exception. [C:\PROJ\FAKE\src\app\Fake.Core.Context\Fake.Core.Context.fsproj]
C:\Users\dragon\.nuget\packages\fsharp.net.sdk\1.0.0-alpha-000007\build\netstandard1.0\FSharp.NET.Core.Sdk.targets(141,9): error MSB3073: The command "dotnet exec --depsfile C:\Users\dragon\.nuget\packages\.tools/dotnet-compile-fsc/1.0.0-preview2-020000/netcoreapp1.0/dotnet-compile-fsc.deps.json --additionalprobingpath C:\Users\dragon\.nuget\packages C:\Users\dragon\.nuget\packages\dotnet-compile-fsc/1.0.0-preview2-020000/lib/netcoreapp1.0/dotnet-compile-fsc.dll @C:\PROJ\FAKE\src\app\Fake.Core.Context\obj\Release\netstandard1.6\netstandard1.6\dotnet-compile.rsp" exited with code 1. [C:\PROJ\FAKE\src\app\Fake.Core.Context\Fake.Core.Context.fsproj]
C:\Users\dragon\AppData\Local\Microsoft\dotnet\sdk\1.0.0-preview4-004233\Sdks\Microsoft.NET.Sdk\build\Microsoft.NET.Sdk.targets(119,5): error : Assets file 'C:\PROJ\FAKE\src\app\Fake.netcore\obj\project.assets.json' doesn't have a target for '.NETCoreApp,Version=v1.0/win10-x64'. Ensure you have restored this project for TargetFramework='netcoreapp1.0' and RuntimeIdentifier='win10-x64'. [C:\PROJ\FAKE\src\app\Fake.netcore\Fake.netcore.fsproj]
```

while 

```
C:\Users\dragon\AppData\Local\Microsoft\dotnet\dotnet.exe publish C:\PROJ\FAKE\src\app\Fake.netcore\Fake.netcore.fsproj --configuration Release --framework "netcoreapp1.0" --output "C:\PROJ\FAKE\nuget\dotnetcore\Fake.netcore\current"
```

is working fine. Is publishing for a specific RID still supported in preview4? I think the problem might be a bug in the paket integration not restoring the correct packages. In preview2 one had to rewrite the `project.json` and do another restore. Is this still true as well?

matthid commented:
Seems like I confused the F# compiler with too much `Environment`. [Is this a known bug](https://github.com/fsharp/FAKE/pull/1281/commits/b34c571d0d43f1c721c93f78da03442afb2bbd7f)?

And the paket dotnet support [is not working on unix](https://travis-ci.org/fsharp/FAKE/builds/189290799#L3102)? /cc @forki

matthid commented:
Seems like something else broke as well :( I have no idea.

smoothdeveloper commented on paket.dependencies:
```diff
@@ -61,19 +62,18 @@ group Build
     nuget FSharp.NET.Sdk ~> 1.0.0 alpha
     //nuget Microsoft.FSharp.Core.netcore 1.0.0-alpha-161023 alpha framework: netstandard1.6, netcoreapp1.0 
-    nuget Microsoft.NET.Sdk 1.0.0-alpha-20161104-2 alpha
+    //nuget Microsoft.NET.Sdk 1.0.0-alpha-20161104-2 alpha
```
isn't the syntax with # ?

matthid replied:
I think both works :)

matthid commented:
@forki Can we drop xunit-1 support? https://www.nuget.org/packages/xunit/1.9.2 

Last version is almost 5 years old and it seems the build server has problems with it (no idea why, maybe F# Core update?).... 
Locally its not running because I would need to install net2/3 which is not installed currently...

forki commented:
I'm fine with dropping xunit 1 support

Am 04.03.2017 14:55 schrieb "Matthias Dittrich" <notifications@github.com>:

> @forki <https://github.com/forki> Can we drop xunit-1 support?
> https://www.nuget.org/packages/xunit/1.9.2
>
> Last version is almost 5 years old and it seems the build server has
> problems with it (no idea why, maybe F# Core update?)....
> Locally its not running because I would need to install net2/3 which is
> not installed currently...
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/fsharp/FAKE/pull/1281#issuecomment-284152840>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AADgNAQPVmWK4rD2RNaSlwD_T1HR6em2ks5riW0-gaJpZM4I-igG>
> .
>

matthid commented:
@forki perfect, than I will update everything to net45 and get rid of it (partly because of https://github.com/fsharp/fsharp/issues/672)

matthid commented:
Now I need to revert https://github.com/fsprojects/Paket/issues/2072 and do that properly ...

dsyme commented:
@matthid How close is this?  People were asking about it at #fsharpX

matthid commented:
@dsyme I know and I'm sorry ;)
But I want to get it kind of right. Next step is to figure out what FCS is trying to tell me and to fix runtime assembly detection in Paket.

I'll be working on this in-depth again next week (at least I hope). And hopefully this will result in some kind of usable realease ...

matthid commented:
obviously no promises and nothing to announce...

matthid commented:
@dsyme figuring out the fsi story on dotnetcore would help me as well. because as it stands now I'll probably go one route with fake and a couple of months later official fsi will go another :(

matthid commented:
https://github.com/fsharp/fsharp/pull/718

matthid commented:
Finally I could get rid of `Microsoft.FSharp.Core.netcore`, but now "sigdata" and "optdata" are no longer published by dnc :(

matthid commented:
Just for you insiders following this thread. This is the version which now bootstrapped FAKE itself https://chocolatey.org/packages/fake/5.0.0-alpha004. You can just 'choco install fake -pre'. Just keep in mind this is alpha. There is pretty much no user experience jet, but it kind of works :)

matthid commented:
By the way its only red on travis because of https://github.com/fsprojects/Paket/issues/2262

matthid commented:
Blocker: https://github.com/fsharp/FSharp.Compiler.Service/issues/763

matthid commented:
https://github.com/dotnet/cli/issues/6465

matthid commented:
Its incredible what changes were necessary throughout the ecosystem to make this possible. When I started it I thought we just integrate paket and things would work :)

After almost a year this is finally green. This means we can finally look at FAKE instead of fixing all the other stuff.

This also means I will now work on branches on top of this and then merge them back here for the time being. Sadly this is less visible for others until we have the access rights figured out.

matthid commented:
https://github.com/fsharp/FAKE/pull/1559
